### PR TITLE
Improve external include & warning support

### LINKED ISSUE #1692: Add support for MSVC external includes
**What problem will this solve?**
premake has support for sysincludedirs for a long time. It works great for clang and gcc in my experience. However for visual studio the directories in sysincludedirs are threated as normal includes. This leads to warning generation in those includes. It's annoying if you want to use the FatalWarnings flag that premake provides.

**What might be a solution?**
Visual Studio has support for external includes since Visual Studio 2017 version 15.6 Preview 1: https://devblogs.microsoft.com/cppblog/broken-warnings-theory/

But premake has no support for it yet. Is it possible to add this feature? If you can point me to the right direction I can implement it myself.

**What other alternatives have you already considered?**
You can work around the missing support by using:
`buildoptions {"/experimental:external  /external:W0"}`
`buildoptions {"/external:I <include path>"}`

The downside of this approach is that the result is not clearly visible in the project settings inside visual studio. It does not show up as  external includes:
external1
but you can see them in the additional options:
external2

### PR DESCRIPTION
**Pull request by englercj**
**What does this PR do?**

Adds support for "external" headers to Visual Studio 2022. This adds three new APIs:

- `externalwarnings` - Controls warning levels for headers considered "external" (`/external:Wn`, `<ExternalWarningLevel>`)
- `externalanglebrackets` - When set to "On" treats all headers included with angle brackets as "external" (`/external:anglebrackets`, `<TreatAngleIncludeAsExternal>`)
- `externalincludedirs` - Include directory paths to treat as "external" (`/external:Ipath`, `<>`)

This is a different take on this support than #1701. Here instead of repurposing `sysincludedirs`, I added a new concept of `externalincludedirs`. I think these two things are fundamentally different as "external" in this context means "not part of my project" which doesn't necessarily mean "system". For example, I use these for some third-party non-system libraries, which I include using double quotes and not angle brackets.

MSVC Docs: https://docs.microsoft.com/en-us/cpp/build/reference/external-external-headers-diagnostics?view=msvc-170

**How does this PR change Premake's behavior?**

Existing behavior should be unaffected, all new functionality is behind new APIs.

**Anything else we should know?**

Add any other context about your changes here.

closes #1692

**Did you check all the boxes?**

- [x] Focus on a single fix or feature; remove any unrelated formatting or code changes
- [x] Add unit tests showing fix or feature works; all tests pass
- [x] Mention any [related issues](https://github.com/premake/premake-core/issues) (put `closes #XXXX` in comment to auto-close issue when PR is merged)
- [x] Follow our [coding conventions](https://github.com/premake/premake-core/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] Minimize the number of commits
- [x] Align [documentation](https://github.com/premake/premake-core/tree/master/website) to your changes

### CODE REVIEW

**Comment by KyrietS** [comment]
Please add your new markdown pages to [website/sidebar.js](https://github.com/premake/premake-core/blob/master/website/sidebars.js)

File: src/tools/msc.lua
Comment (Jarod42) [inline src/tools/msc.lua:412]: To be coherent with other toolset, that implementation should be included in `getincludedirs`
and just call `msc.getincludedirs(cfg, nil, nil, cfg.externalincludedirs, nil)`

**Comment by nickclark2016** [comment]
Out of scope of this PR, but how exactly should this be handled by other build systems?  For build systems without this concept, should `externalincludedirs` be treated as `sysincludedirs`?  I think this question should be answered and the consequences understood before we approve/deny this.

**Comment by Jarod42** [comment]
> I think these two things are fundamentally different as "external" in this context means "not part of my project" which doesn't necessarily mean "system".

Note that msvc does nothing for `sysincludedirs`, so it might just be a naming issue.
What would you place in "system"?
Aliasing might be an option if really you dislike the name.

**Comment by StellaASchlotter** [comment]
I am the author of https://github.com/premake/premake-core/pull/1701. Unfortunately I can't continue working on premake for the next few months. However I want to point out that I think it is a bug in premake that sysincluddirs do not supress warnings in visual studio  but do in gcc and clang. I wanted to address this as part of my PR. 

My idea was to make sysincludedirs behave the same on all systems so I don't need multiple instructions for the same behavior.

But I am okay with closing my PR and going forward with this one. Any support for externalincludedirs is better than none.

**Comment by englercj** [comment]
> What would you place in "system"?

I guess I'd separate include paths I expect to exist on the build system (Windows SDK or glibc++, for example). And external would be dependencies of my application that aren't from the system, for example sqlite, freetype, or other third-party libraries that are external to my app but not part of the system.

Maybe this distinction is moot or undesired, and that's fine I'm open to any solution that lets me tag file paths as external in MSVC so I can disable warnings when including them.

> My idea was to make sysincludedirs behave the same on all systems so I don't need multiple instructions for the same behavior.

If I change this PR to fall back to sysincludedirs on other build systems then you will get the behavior you expect of this new instruction silencing warnings on all platforms. Is that a reasonable solution?

File: src/tools/msc.lua
Comment (englercj) [reply src/tools/msc.lua:412]: Will fix.

File: website/docs/externalincludedirs.md
Comment (englercj) [reply website/docs/externalincludedirs.md:36]: Will fix.

File: website/docs/externalwarnings.md
Comment (englercj) [reply website/docs/externalwarnings.md:37]: Will fix.

File: website/docs/externalwarnings.md
Comment (englercj) [reply website/docs/externalwarnings.md:1]: I copy-pasted this from `warnings.md`, did you want me to change the word there too?

File: website/docs/externalincludedirs.md
Comment (englercj) [reply website/docs/externalincludedirs.md:1]: Will fix.

File: website/docs/externalanglebrackets.md
Comment (englercj) [reply website/docs/externalanglebrackets.md:1]: Will fix.

**Comment by nickclark2016** [comment]
I've had some time to dig into the implications of this as opposed to co-opting the existing API.  As it currently stands, I don't think I'd want to accept the PR as-is (however I appreciate the work 😃).  Below I've listed some of my research into the issue as well as what my take on the correct course of action is.

* `externalincludedirs` - This is fundamentally no different than `sysincludedirs`.  While this may not semantically be what you want, they have roughly the same meaning.  These both seem to mean any include directory other than includes from projects that the user owns.
* `externalanglebrackets` - This is something unique to Visual Studio as far as I can tell.  Since this has no analog to any other toolchain, should we make the API part of the visual studio module or should we ignore it in other toolchains.
* `externalwarnings` - This equates to the `-Wsystem-headers` flag in GCC/Clang.

What do I propose?

* I believe that we should coalesce the existing `sysincludedirs` and `externalincludedirs` to a single API called `externalincludedirs`.  I think this name works better, as it is the more all-encompassing of the two.  This will prevent the need for an exporter-specific API.  This would behave by writing `externalincludedirs` to the `externalincludedirs` in Visual Studio, but for GNU makefile it would write it to `-isystem`.
* For `externalwarnings`, I believe we should try to mirror as closely to the current `warnings` options as possible.  See below table for flag mapping.

| Premake Symbol Value | Visual Studio | GNU Makefile     |
|----------------------|---------------|------------------|
| Off                  | /external:W0  |                  |
| Default              | /external:W3  | -Wsystem-headers |
| High                 | /external:W4  | -Wsystem-headers |
| Extra                | /external:W4  | -Wsystem-headers |
| Everything           | /external:W4  | -Wsystem-headers |

* Finally, I believe that the `externalanglebrackets` API should be specific to Visual Studio.

This would have impacts on the API that already exists for `sysincludedirs`.  Instead of removing the API, I would have it deprecated in favor of the new `externalincludedirs` and removed in the full release.

Thoughts on this?

**Comment by starkos** [comment]
+1 on this plan.

> This would have impacts on the API that already exists for sysincludedirs. Instead of removing the API, I would have it deprecated in favor of the new externalincludedirs and removed in the full release.

In favor of deprecating, but would probably not remove until v6 at this point, IMHO. Another option is to make `sysincludedirs` and alias for `externalincludedirs`?

**Comment by englercj** [comment]
> I believe that we should coalesce the existing sysincludedirs and externalincludedirs to a single API called externalincludedirs

I think this makes a lot of sense, I'll get that done.

> For externalwarnings, I believe we should try to mirror as closely to the current warnings options as possible. See below table for flag mapping.

I can make this change, but the reason I had them different originally is because I didn't want the same words with different meanings. For example, `"Everything"` in the `warnings` API means `/Wall` but in the `externalwarnings` it would mean `/W4`. That may be surprising for some users so instead of having multiple `/W4`s I just cut that one out. I'll add it back in for now and let people review.

> Finally, I believe that the externalanglebrackets API should be specific to Visual Studio.

It is currently only used in vs project generation and ignored by other toolsets. Did you want me to do something different here?

**Comment by englercj** [comment]
> In favor of deprecating, but would probably not remove until v6 at this point, IMHO. Another option is to make sysincludedirs and alias for externalincludedirs?

I don't have strong feelings on this, ahppy to do whatever people want here.

**Comment by nickclark2016** [comment]
> > I believe that we should coalesce the existing sysincludedirs and externalincludedirs to a single API called externalincludedirs
> 
> I think this makes a lot of sense, I'll get that done.
> 
> > For externalwarnings, I believe we should try to mirror as closely to the current warnings options as possible. See below table for flag mapping.
> 
> I can make this change, but the reason I had them different originally is because I didn't want the same words with different meanings. For example, `"Everything"` in the `warnings` API means `/Wall` but in the `externalwarnings` it would mean `/W4`. That may be surprising for some users so instead of having multiple `/W4`s I just cut that one out. I'll add it back in for now and let people review.
> 
> I also don't like `"Default"` being `/external:W3`. The visual studio default is to inherit the project's warning level, which matches the behavior of older versions. I'd like to keep that behavior as the default.

For this, I think we should personally make the `warnings` and `externalwarnings` flags match as much as possible. To me, this would be the most straight forward.  It would be nice to have `inherit` as well, but I think the mismatch for `Default` would be confusing.

> > Finally, I believe that the externalanglebrackets API should be specific to Visual Studio.
> 
> It is currently only used in vs project generation and ignored by other toolsets. Did you want me to do something different here?

So that's one option. Another option would be to register the API on the vstudio module.  See this file: https://github.com/premake/premake-core/blob/master/modules/vstudio/_preload.lua

**Comment by englercj** [comment]
OK I think I've addressed all the feedback at this point, aside from deprecating `sysincludedirs` and making it an alias of `externalincludedirs`. I wanted the naming debate of which one to keep settle first so I knew how to proceed.

File: src/tools/msc.lua
Comment (Jarod42) [inline src/tools/msc.lua:258]: Logically, test should be against the toolset version, as other actions can use msc toolset.

File: website/docs/externalanglebrackets.md
Comment (Jarod42) [inline website/docs/externalanglebrackets.md:23]:
```suggestion
Visual Studio 2022 version 17.0 or later.
```
typo has been copy pasted...

File: website/docs/externalwarnings.md
Comment (Jarod42) [inline website/docs/externalwarnings.md:19]: To document.

**Comment by Jarod42** [comment]
My opinion about naming:
- `externalincludedirs` seems clearer than `sysincludedirs`.
- Having both, one as aliasing of the other is ok.
- `sysincludedirs` has historic advantage, but deprecating `sysincludirs` is ok for me.

If we keep only `sysincludedirs`, then the new `externalanglebrackets`, `externalwarnings` should be renamed IMO.

File: src/tools/msc.lua
Comment (englercj) [reply src/tools/msc.lua:258]: I was thinking that as well. How do I check that here?

**Comment by englercj** [comment]
Made changes to deprecate `sysincluddirs` and only have an implementation for `externalincludedirs`.

File: src/_premake_init.lua
Comment (Jarod42) [inline src/_premake_init.lua:1479]: Should it be `api.alias()` instead?

File: website/docs/externalincludedirs.md
Comment (Jarod42) [inline website/docs/externalincludedirs.md:9]:
```suggestion
Include files located via an external include directory are treated specially, see [externalwarnings](externalwarnings.md).
```

File: website/docs/sysincludedirs.md
Comment (Jarod42) [inline website/docs/sysincludedirs.md:37]: Page should probably just state, alias of `[externalincludedirs](externalincludedirs.md)` and deprecated statut.

File: modules/vstudio/vs2010_vcxproj.lua
Comment (Jarod42) [inline modules/vstudio/vs2010_vcxproj.lua:2151]: Those 2 methods should probably be merged, as they both handle cfg.externalincludedirs.
or vs2022_vcxproj created?

File: modules/vstudio/vs2010_vcxproj.lua
Comment (Jarod42) [inline modules/vstudio/vs2010_vcxproj.lua:2857]: So probably dedicated file for vs2022 stuff.

File: src/_premake_init.lua
Comment (englercj) [reply src/_premake_init.lua:1479]: I copied the pattern of `debuggerflavor` being deprecated in `vstudio/_preload.lua`. Let me know if I should do something different.

File: modules/vstudio/vs2010_vcxproj.lua
Comment (englercj) [reply modules/vstudio/vs2010_vcxproj.lua:2857]: I combined the functions, moving this into a vcxproj file specifically for 2022 would be a different pattern than all the other vs-version specific stuff in this file which checks _ACTION.

**Comment by englercj** [comment]
Are there additional changes people are interested in seeing here?

File: src/tools/msc.lua
Comment (Jarod42) [reply src/tools/msc.lua:259]: I would say something like `cfg.toolset >= "msc-v143"`.

Comment (Jarod42) [reply src/tools/msc.lua:259]: @starkos: from [toolset](https://premake.github.io/docs/toolset/), it seems we can use `"v143"` directly instead of `"msc-v143"`, I didn't see code to unify it afterward.
As `"vyyy" > "msc-xxx"`, that check might be problematic. which is the correct method?

**Review by starkos — APPROVED** [review]
This looks good to me now. I've updated the PR title to be more accurate, and approving for merge.

**Comment by starkos** [comment]
@englercj Can you please rebase this against the latest main branch, and squash the commits?

**Comment by Jarod42** [comment]
@starkos:
It seems you miss my comment:
> from toolset, it seems we can use "v143" directly instead of "msc-v143", I didn't see code to unify it afterward.
As "vyyy" > "msc-xxx", that check might be problematic. which is the correct method?

**Comment by starkos** [comment]
> @starkos: It seems you miss my comment:
> > from toolset, it seems we can use "v143" directly instead of "msc-v143", I didn't see code to unify it afterward.
> > As "vyyy" > "msc-xxx", that check might be problematic. which is the correct method?

I did miss it, thanks for escalating. It [looks like the value will get normalized to "msc-v143"](https://github.com/premake/premake-core/blob/master/src/base/tools.lua#L32-L53).

File: src/tools/msc.lua
Comment (Jarod42) [reply src/tools/msc.lua:259]: `cfg.toolset >= "msc-v143"` seems the way to go.

Comment (englercj) [reply src/tools/msc.lua:259]: Changing this line to `if cfg.toolset >= "msc-v143" then` causes an error in the tests:

```
[  FAILED  ] 1 tests, listed below:
[  FAILED  ] tools_msc.cflags_onExternalIncludeDirs
[string "src/tools/msc.lua"]:187: attempt to compare string with nil
stack traceback:
        [string "src/tools/msc.lua"]:187: in field 'getincludedirs'
        ...s/engle/source/oss/premake-core/tests/tools/test_msc.lua:301: in function 'testFunction'
```

Seems like the `cfg.toolset` value is not always set.

**Comment by englercj** [comment]
Hey @starkos, just wanted to ping about this comment: https://github.com/premake/premake-core/pull/1754#discussion_r758605807

I tried the method of comparing `cfg.toolset` but it is sometimes `nil`. Please advise.

**Comment by Jarod42** [comment]
I suppose `p.action.set("vs2010")` doesn't set associated toolset.
You probably need to set toolset also afterward :-/

I hope it is only a Unit testing issue.

File: src/tools/msc.lua
Comment (Jarod42) [reply src/tools/msc.lua:259]: `cfg.toolset and cfg.toolset >= "msc-v143"` here and set toolset or action in test_msc.lua does the job.
Actually, in UT of toolset, we don't set neither action nor toolset, that'is why you got issue.

**Comment by englercj** [comment]
Rebased and ready for merge.

File: src/tools/msc.lua
Comment (englercj) [reply src/tools/msc.lua]: Fixed.

**Comment by englercj** [comment]
Fixed the toolset check and squashed the commits onto latest master. I think we're good to go.

**Review by Jarod42 — APPROVED** [review]

**Comment by peter1745** [comment]
Can't wait for this to be merged in so vs2022 support isn't *meh* anymore

**Comment by englercj** [comment]
Rebased atop latest master, ready for merge AFAIK.

**Comment by starkos** [comment]
Done, thanks!

**Comment by StellaASchlotter** [comment]
@englercj was there a reason why its only for visual studio 2022 and up? Externalincludes are supported in vs2019.

**Comment by englercj** [comment]
@StellaASchlotter It was experimental in most versions of vs2019 and required the `/experimental:external` flag to be set in order to work. It was only non-experimental as of vs2019 16.10. I didn't deal with the complexity of checking when to set the experimental flag because vs2022 support was sufficient for my needs, and it is guaranteed to work there in all versions.

**Comment by StellaASchlotter** [comment]
Ah thanks for this information. I was not aware of this. I will upgrade the VS version in my team then.

**Pull request merged**
